File: server/test/EnvelopeTest.kt
package ghostbe.server

import kotlin.test.Test
import kotlin.test.assertEquals

class EnvelopeTest {
    @Test
    fun `decodes a request envelope from json`() {
        val json = """
            {"method":"GET","url":"https://api.example.com/v1/users/42?active=true","headers":{"Accept":"application/json"},"body":null}
        """.trimIndent()
        val envelope = RequestEnvelope.fromJson(json)
        assertEquals("GET", envelope.method)
        assertEquals("https://api.example.com/v1/users/42?active=true", envelope.url)
        assertEquals(mapOf("Accept" to "application/json"), envelope.headers)
        assertEquals(null, envelope.body)
    }

    @Test
    fun `encodes a mock response envelope to json`() {
        val envelope: ResponseEnvelope = ResponseEnvelope.Mock(
            status = 200,
            headers = mapOf("Content-Type" to "application/json"),
            body = "eyJvayI6dHJ1ZX0="
        )
        val json = envelope.toJson()
        assertEquals(true, json.contains("\"intercept\":true"))
        assertEquals(true, json.contains("\"status\":200"))
    }

    @Test
    fun `encodes a passthrough envelope to json`() {
        val envelope: ResponseEnvelope = ResponseEnvelope.Passthrough()
        assertEquals("""{"intercept":false}""", envelope.toJson())
    }
}

File: client-flutter/test/envelope_test.dart
import 'dart:convert';
import 'package:test/test.dart';
import 'package:ghost_be/src/envelope.dart';

void main() {
  test('RequestEnvelope encodes to the exact wire JSON shape', () {
    final envelope = RequestEnvelope(
      method: 'GET',
      url: 'http://x/v1/users/42',
      headers: {'Accept': 'application/json'},
      body: null,
    );
    final decoded = jsonDecode(jsonEncode(envelope.toJson())) as Map<String, dynamic>;
    expect(decoded, {
      'method': 'GET',
      'url': 'http://x/v1/users/42',
      'headers': {'Accept': 'application/json'},
      'body': null,
    });
  });

  test('ResponseEnvelope.fromJson decodes a mock response', () {
    final json = {
      'intercept': true,
      'status': 201,
      'headers': {'Content-Type': 'application/json'},
      'body': 'eyJvayI6dHJ1ZX0=',
    };
    final decision = ResponseEnvelope.fromJson(json);
    expect(decision, isA<MockResponseEnvelope>());
    final mock = decision as MockResponseEnvelope;
    expect(mock.status, 201);
    expect(mock.headers, {'Content-Type': 'application/json'});
    expect(mock.body, 'eyJvayI6dHJ1ZX0=');
  });

  test('ResponseEnvelope.fromJson decodes a passthrough response', () {
    final decision = ResponseEnvelope.fromJson({'intercept': false});
    expect(decision, isA<PassthroughResponseEnvelope>());
  });

  test('ResponseEnvelope.fromJson treats a missing intercept key as passthrough', () {
    final decision = ResponseEnvelope.fromJson({});
    expect(decision, isA<PassthroughResponseEnvelope>());
  });
}
